src/shipit_code_coverage/tests/test_utils.py

```diff
@@ -57,12 +58,32 @@ def test_threadpoolexecutorresult():
     except:
         assert True
 
+    # Tast that ThreadPoolExecutorResult's context returns as soon as a task fails.
     try:
         now = time.time()
         with utils.ThreadPoolExecutorResult() as executor:
-            executor.submit(lambda: time.sleep(10))
+            executor.submit(lambda: time.sleep(5))
             executor.submit(do_raise)
         assert False
     except:
         assert True
-        assert time.time() - now < 5
+        assert time.time() - now < 2
+
+    # Test that futures that were not scheduled yet are cancelled.
+    try:
+        with utils.ThreadPoolExecutorResult(max_workers=1) as executor:
+            f1 = executor.submit(lambda: time.sleep(1))
+            f2 = executor.submit(do_raise)
+            executor.submit(lambda: time.sleep(1))
+            f4 = executor.submit(lambda: time.sleep(1))
+            f5 = executor.submit(lambda: time.sleep(1))
+            f6 = executor.submit(lambda: time.sleep(1))
+        assert False
+    except:
+        assert True
+        assert f1.done() and not f1.cancelled()
+        assert f2.exception() is not None
+        # Not enough time to cancel the third feature, scheduled right after the one which raises an exception.
```

Comment:
I don't understand this. Could you please elaborate why the third feature isn't cancelled in time?

Reply:
My hunch is that the ThreadPoolExecutor is scheduling the third future to run before we have a chance to cancel it (futures that are already running can't be cancelled), as soon as the second finishes.

Reply:
I've added a more complete comment.